tpm2/src/util/nsl_util.c

```diff
@@ -0,0 +1,983 @@
+    return 1;
+  }
+
+  // Determine the length of the plaintext buffer.
+  result = EVP_PKEY_decrypt(ctx, NULL, p_len, c, c_len);
+  if (result <= 0)
+  {
+    kmyth_log(LOG_ERR,
+              "Failed to determine the length of the plaintext buffer.");
+    return 1;
+  }
+
+  // Allocate the plaintext buffer.
+  *p = calloc(*p_len, sizeof(unsigned char));
+  if (*p == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the plaintext buffer.");
+    return 1;
+  }
+
+  // Decrypt the ciphertext.
+  result = EVP_PKEY_decrypt(ctx, *p, p_len, c, c_len);
+  if (result <= 0)
+  {
+    kmyth_log(LOG_ERR, "Failed to decrypt the ciphertext.");
+    return 1;
+  }
+
+  return 0;
+}
+
+//
+// build_nonce_request()
+// 
+int build_nonce_request(EVP_PKEY_CTX * ctx,
+                        unsigned char *nonce, size_t nonce_len,
+                        unsigned char *id, size_t id_len,
+                        unsigned char **request, size_t *request_len)
+{
+  // Allocate the unencrypted request buffer.
+  unsigned char *message = NULL;
+  size_t message_len = id_len + nonce_len + (2 * sizeof(size_t));
+
+  message = calloc(message_len, sizeof(unsigned char));
+  if (message == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the message buffer.");
+    return 1;
+  }
+
+  // Build the nonce request message.
+  unsigned char *index = message;
+
+  memcpy(index, &nonce_len, sizeof(size_t));
+  index += sizeof(size_t);
+  memcpy(index, nonce, nonce_len);
+  index += nonce_len;
+  memcpy(index, &id_len, sizeof(size_t));
+  index += sizeof(size_t);
+  memcpy(index, id, id_len);
+
+  // Encrypt the nonce request and then clean up the unencrypted request.
+  int result =
+    encrypt_with_key_pair(ctx, message, message_len, request, request_len);
+
+  memset(message, 0, message_len);
+  free(message);
+
+  // Handle encryption errors if any occurred.
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to encrypt the nonce request.");
+    return 1;
+  }
+
+  return 0;
+}
+
+//
+// parse_nonce_request()
+//
+int parse_nonce_request(EVP_PKEY_CTX * ctx,
+                        unsigned char *request, size_t request_len,
+                        unsigned char **nonce, size_t *nonce_len,
+                        unsigned char **id, size_t *id_len)
+{
+  // Decrypt the nonce request.
+  unsigned char *message = NULL;
+  size_t message_len = 0;
+  int result =
+    decrypt_with_key_pair(ctx, request, request_len, &message, &message_len);
+
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to decrypt the nonce request.");
+    return 1;
+  }
+  unsigned char *index = message;
+
+  // Parse out the nonce.
+  memcpy(nonce_len, index, sizeof(size_t));
+  index += sizeof(size_t);
+  *nonce = calloc(*nonce_len, sizeof(unsigned char));
+  if (*nonce == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the nonce buffer.");
+
+    *nonce_len = 0;
+
+    memset(message, 0, message_len);
```

Comment:
We should probably use secure_memset() (in tpm2/src/util/memory_util.c), which prevents against potential compiler optimization issues.

Reply:
Done